presto-native-execution/presto_cpp/main/PrestoExchangeSource.h

```diff
@@ -23,43 +23,46 @@
 
 namespace facebook::presto {
 
-class Backoff {
+class PrestoExchangeSource : public velox::exec::ExchangeSource {
  public:
-  Backoff(int64_t maxWaitMs = 1000)
-      : maxWaitMs_(maxWaitMs), startMs_(velox::getCurrentTimeMs()) {}
-
-  // Returns the delay in millis to wait before next try. This is an exponential
-  // backoff delay with jitter. The first call to this always returns 0.
-  int64_t nextDelayMs() {
-    ++numTries_;
-    if (numTries_ == 1) {
-      return 0;
+  class RetryState {
+   public:
+    RetryState(int64_t maxWaitMs = 1000)
+        : maxWaitMs_(maxWaitMs), startMs_(velox::getCurrentTimeMs()) {}
+
+    // Returns the delay in millis to wait before next try. This is an
+    // exponential backoff delay with jitter. The first call to this always
+    // returns 0.
+    int64_t nextDelayMs() {
+      if (++numTries_ == 1) {
+        return 0;
+      }
+      auto rng = folly::ThreadLocalPRNG();
+      return folly::futures::detail::retryingJitteredExponentialBackoffDur(
+                 numTries_ - 1,
+                 std::chrono::milliseconds(kMinBackoffMs),
+                 std::chrono::milliseconds(kMaxBackoffMs),
+                 kJitterParam,
+                 rng)
+          .count();
+    }
+
+    // Returns whether we have exhausted all retries. We only retry if we spent
+    // less than maxWaitMs_ time after we first started.
+    bool isExhausted() const {
+      return velox::getCurrentTimeMs() - startMs_ > maxWaitMs_;
     }
-    auto rng = folly::ThreadLocalPRNG();
+
+   private:
+    int64_t maxWaitMs_;
```

Comment:
Not marked as const, because copy constructor is deleted when some members are const.

Reply:
Do we need to copy retryState?

Reply:
I set it every time a new request is to be made. It will be friendlier to mutable configs this way.